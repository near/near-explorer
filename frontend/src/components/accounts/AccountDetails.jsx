import { Row, Col } from "react-bootstrap";

import Balance from "../utils/Balance";

import AccountCard from "./AccountCard";

export default ({ account }) => {
  return (
    <>
      <Row noGutters="true">
        <Col className="account-info-container">
          <Row noGutters="true">
            <Col md="3">
              <AccountCard
                title="â“ƒ Balance"
                text={<Balance amount={account.amount} />}
                cls="account-card-no-side-border"
              />
            </Col>
            <Col md="3">
              <AccountCard
                title="Transactions"
                imgLink="/static/images/icon-m-transaction.svg"
                text={
                  <>
                    <span>
                      &uarr;{account.outTransactionsCount.toLocaleString()}
                    </span>
                    &nbsp;&nbsp;
                    <span>
                      &darr;{account.inTransactionsCount.toLocaleString()}
                    </span>
                  </>
                }
              />
            </Col>
            <Col md="3">
              <AccountCard
                title="Staked"
                imgLink="/static/images/icon-m-filter.svg"
                text={<Balance amount={account.staked} />}
              />
            </Col>
            <Col md="3">
              <AccountCard
                title="Storage (Used / Paid)"
                imgLink="/static/images/icon-m-block.svg"
                text={`${account.storage_usage.toLocaleString()} / ${account.storage_paid_at.toLocaleString()}`}
              />
            </Col>
          </Row>
        </Col>
      </Row>
      <style jsx global>{`
        .account-card-created-text {
          font-size: 18px;
          font-weight: 500;
          color: #4a4f54;
        }

        .account-card-author-text {
          color: #0072ce;
        }

        .account-card-parent-hash-text {
          font-size: 18px;
          font-weight: 500;
          color: #6ad1e3;
          cursor: pointer;
          text-decoration: none !important;
        }

        .account-card-parent-hash-text:hover {
          color: #6ad1e3;
        }

        .account-card-parent-hash {
          background-color: #f8f8f8;
        }

        .account-info-container .account-card {
          border-top: 0;
          border-bottom-width: 2px;
          border-right: 0;
          border-left-width: 2px;
        }

        .account-info-container .account-card-no-side-border {
          border-left: 0;
          border-right: 0;
        }

        .account-info-container .account-card-no-border {
          border: 0;
        }

        .account-info-container {
          border: solid 4px #e6e6e6;
          border-radius: 4px;
        }
      `}</style>
    </>
  );
};
