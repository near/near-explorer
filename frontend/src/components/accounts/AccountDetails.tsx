import moment from "moment";

import React from "react";

import { Row, Col } from "react-bootstrap";

import * as A from "../../libraries/explorer-wamp/accounts";
import BlocksApi from "../../libraries/explorer-wamp/blocks";

import Balance from "../utils/Balance";
import CardCell from "../utils/CardCell";
import TransactionLink from "../utils/TransactionLink";
import BlockLink from "../utils/BlockLink";

export interface Props {
  account: A.Account;
}

interface State {
  storagePaidAtBlockHash: string | null;
}

export default class extends React.Component<Props, State> {
  state: State = {
    storagePaidAtBlockHash: null
  };

  fetchBlockHash = async (height: string) => {
    new BlocksApi()
      .getBlockInfo(height)
      .then(block => this.setState({ storagePaidAtBlockHash: block.hash }))
      .catch(err => console.error(err));
  };

  componentDidUpdate(preProps: Props) {
    if (this.props.account !== preProps.account) {
      this.fetchBlockHash(this.props.account.storagePaidAt.toString());
    }
  }
  componentDidMount() {
    this.fetchBlockHash(this.props.account.storagePaidAt.toString());
  }

  render() {
    const { account } = this.props;
    const { storagePaidAtBlockHash } = this.state;
    return (
      <div className="account-info-container">
        <Row noGutters>
          <Col md="2">
            <CardCell
              title="â“ƒ Balance"
              text={<Balance amount={account.amount} />}
              className="border-0"
            />
          </Col>
          <Col md="3">
            <CardCell
              title="Transactions"
              imgLink="/static/images/icon-m-transaction.svg"
              text={
                <>
                  <span>
                    &uarr;{account.outTransactionsCount.toLocaleString()}
                  </span>
                  &nbsp;&nbsp;
                  <span>
                    &darr;{account.inTransactionsCount.toLocaleString()}
                  </span>
                </>
              }
            />
          </Col>
          <Col md="2">
            <CardCell
              title="Locked"
              imgLink="/static/images/icon-m-filter.svg"
              text={<Balance amount={account.locked} />}
            />
          </Col>
          <Col md="3">
            <CardCell
              title="Storage Used (Bytes)"
              imgLink="/static/images/icon-storage.svg"
              text={`${account.storageUsage.toLocaleString()} B`}
            />
          </Col>
          <Col md="2">
            <CardCell
              title="Last Paid"
              imgLink="/static/images/icon-m-block.svg"
              text={
                storagePaidAtBlockHash ? (
                  <BlockLink blockHash={storagePaidAtBlockHash}>
                    {`#${account.storagePaidAt.toLocaleString()}`}
                  </BlockLink>
                ) : (
                  `#${account.storagePaidAt.toLocaleString()}`
                )
              }
            />
          </Col>
        </Row>
        <Row noGutters className="border-0">
          <Col md="5">
            <CardCell
              title="Created"
              text={
                typeof account.timestamp === "number"
                  ? moment(account.timestamp).format(
                      "MMMM DD, YYYY [at] h:mm:ssa"
                    )
                  : account.timestamp
              }
              className="block-card-created account-card-back border-0"
            />
          </Col>
          <Col md="7">
            <CardCell
              title="Creation Hash"
              text={
                account.address === "" ? (
                  ""
                ) : (
                  <TransactionLink transactionHash={account.address}>
                    {account.address}
                  </TransactionLink>
                )
              }
              className="block-card-created-text account-card-back border-0"
            />
          </Col>
        </Row>
        <style jsx global>{`
          .account-info-container {
            border: solid 4px #e6e6e6;
            border-radius: 4px;
          }

          .account-info-container > .row:first-of-type .card-cell-text {
            font-size: 24px;
          }

          .account-info-container > .row {
            border-bottom: 2px solid #e6e6e6;
          }

          .account-card-back {
            background-color: #f8f8f8;
          }
        `}</style>
      </div>
    );
  }
}
