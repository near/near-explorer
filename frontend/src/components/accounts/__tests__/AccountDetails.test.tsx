import renderer from "react-test-renderer";

import AccountDetails from "../AccountDetails";

describe("<AccountDetails />", () => {
  it("renders", () => {
    expect(
      renderer.create(
        <AccountDetails
          account={{
            id: "account",
            amount: "1",
            locked: "1",
            storageUsage: 0,
            storagePaidAt: 1,
            inTransactionsCount: 0,
            outTransactionsCount: 0,
            timestamp: Number(new Date(2019, 1, 1)),
            address: "EVvWW1S9BFaEjY1JBNSdstb7ZTtTFjQ6cygkbw1KY4tL"
          }}
        />
      )
    ).toMatchSnapshot();
  });
});
