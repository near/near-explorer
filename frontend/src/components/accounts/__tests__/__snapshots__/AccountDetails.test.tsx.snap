// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AccountDetails /> renders 1`] = `
<div
  className="jsx-1455372711 account-info-container"
>
  <div
    className="row no-gutters"
  >
    <div
      className="col-md-3"
    >
      <div
        className="card-cell border-0 card"
      >
        <div
          className="card-body"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="card-cell-title align-self-center col-md-12 col-auto"
            >
              Ⓝ Balance
              <div
                className="jsx-1104606862 term-helper"
              >
                <img
                  className="jsx-1104606862 info"
                  onClick={[Function]}
                  src="/static/images/icon-info.svg"
                />
              </div>
            </div>
            <div
              className="ml-auto card-cell-text align-self-center col-md-12 col-auto"
            >
              <span
                onBlur={[Function]}
                onFocus={[Function]}
                onMouseOut={[Function]}
                onMouseOver={[Function]}
              >
                &lt;0.00001
                 Ⓝ
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="col-md-3"
    >
      <div
        className="card-cell  card"
      >
        <div
          className="card-body"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="card-cell-title align-self-center col-md-12 col-auto"
            >
              Ⓝ Locked
              <div
                className="jsx-1104606862 term-helper"
              >
                <img
                  className="jsx-1104606862 info"
                  onClick={[Function]}
                  src="/static/images/icon-info.svg"
                />
              </div>
            </div>
            <div
              className="ml-auto card-cell-text align-self-center col-md-12 col-auto"
            >
              <span
                onBlur={[Function]}
                onFocus={[Function]}
                onMouseOut={[Function]}
                onMouseOver={[Function]}
              >
                &lt;0.00001
                 Ⓝ
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="col-md-3"
    >
      <div
        className="card-cell  card"
      >
        <div
          className="card-body"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="card-cell-title align-self-center col-md-12 col-auto"
            >
              <img
                className="jsx-2893410015 card-cell-title-img"
                src="/static/images/icon-m-transaction.svg"
              />
              Transactions
              <div
                className="jsx-1104606862 term-helper"
              >
                <img
                  className="jsx-1104606862 info"
                  onClick={[Function]}
                  src="/static/images/icon-info.svg"
                />
              </div>
            </div>
            <div
              className="ml-auto card-cell-text align-self-center col-md-12 col-auto"
            >
              <span
                className="jsx-1455372711"
              >
                ↑
                0
              </span>
                
              <span
                className="jsx-1455372711"
              >
                ↓
                0
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="col-md-3"
    >
      <div
        className="card-cell  card"
      >
        <div
          className="card-body"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="card-cell-title align-self-center col-md-12 col-auto"
            >
              <img
                className="jsx-2893410015 card-cell-title-img"
                src="/static/images/icon-storage.svg"
              />
              Storage Used
              <div
                className="jsx-1104606862 term-helper"
              >
                <img
                  className="jsx-1104606862 info"
                  onClick={[Function]}
                  src="/static/images/icon-info.svg"
                />
              </div>
            </div>
            <div
              className="ml-auto card-cell-text align-self-center col-md-12 col-auto"
            >
              0 B
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="border-0 row no-gutters"
  >
    <div
      className="col-md-4"
    >
      <div
        className="card-cell block-card-created account-card-back border-0 card"
      >
        <div
          className="card-body"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="card-cell-title align-self-center col-md-12 col-auto"
            >
              Created
              <div
                className="jsx-1104606862 term-helper"
              >
                <img
                  className="jsx-1104606862 info"
                  onClick={[Function]}
                  src="/static/images/icon-info.svg"
                />
              </div>
            </div>
            <div
              className="ml-auto card-cell-text align-self-center col-md-12 col-auto"
            >
              February 01, 2019 at 12:00:00am
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="col-md-8"
    >
      <div
        className="card-cell block-card-created-text account-card-back border-0 card"
      >
        <div
          className="card-body"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="card-cell-title align-self-center col-md-12 col-auto"
            >
              Creation Hash
              <div
                className="jsx-1104606862 term-helper"
              >
                <img
                  className="jsx-1104606862 info"
                  onClick={[Function]}
                  src="/static/images/icon-info.svg"
                />
              </div>
            </div>
            <div
              className="ml-auto card-cell-text align-self-center col-md-12 col-auto"
            >
              <a
                className="transaction-link"
                href="/transactions/EVvWW1S9BFaEjY1JBNSdstb7ZTtTFjQ6cygkbw1KY4tL"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                EVvWW1S9BFaEjY1JBNSdstb7ZTtTFjQ6cygkbw1KY4tL
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
